File: leetcode/C++/Path_sumII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>>res;
        vector<int>r1;
        Sum(root, sum, res, r1);
        
        return res;
    }
    
    void Sum(TreeNode *root, int sum, vector<vector<int> > &res, vector<int>&r1)
    {
        if(!root) return ;
        r1.push_back(root->val);
        if(root->val == sum && root->left == NULL && root->right == NULL)
            res.push_back(r1);
        Sum(root->left, sum-(root->val), res, r1);
        Sum(root->right, sum-(root->val), res, r1);
        r1.pop_back();
    }
};

File: lintcode/divide_conquer&tree/path_sum_ii.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) [][]int {
	var res [][]int
	if root == nil{
		return res
	}
	if root.Left == nil && root.Right == nil{
		if sum == root.Val{
			res = [][]int{[]int{root.Val}}
		}
		return res
	}
	left := pathSum(root.Left, sum-root.Val)
	right := pathSum(root.Right, sum-root.Val)
	for _, v := range left{
		v = append(v[:1], v[0:]...)
		v[0] = root.Val
		res = append(res, v)
	}
	for _, v := range right{
		v = append(v[:1], v[0:]...)
		v[0] = root.Val
		res = append(res, v)
	}
	return res
}

